manager/controlapi/ca_rotation.go

```diff
@@ -0,0 +1,345 @@
+			return grpc.Errorf(codes.InvalidArgument, "must specify CA certificate for each external CA")
+		}
+	}
+
+	if needsExternalCAs(cluster.RootCA) && !hasAtLeastOneExternalCA(ctx, config.ExternalCAs, securityConfig, cluster.RootCA.CACert) {
+		return grpc.Errorf(codes.InvalidArgument, "there must be at least one valid, reachable external CA corresponding to the current CA certificate")
+	}
+
+	if len(config.SigningCACert) > 0 { // a signing cert is specified
+		if needsExternalCAs(config) && !hasAtLeastOneExternalCA(ctx, config.ExternalCAs, securityConfig, config.SigningCACert) {
+			return grpc.Errorf(codes.InvalidArgument, "there must be at least one valid, reachable external CA corresponding to the desired CA certificate")
+		}
+	} else if config.ForceRotate == cluster.RootCA.LastForcedRotation && cluster.RootCA.RootRotation != nil {
+		// no cert is specified but force rotation hasn't changed (so we are happy with the current configuration) and there's an outstanding root rotation
+		if needsExternalCAs(cluster.RootCA.RootRotation) && !hasAtLeastOneExternalCA(ctx, config.ExternalCAs, securityConfig, cluster.RootCA.RootRotation.CACert) {
+			return grpc.Errorf(codes.InvalidArgument, "there must be at least one valid, reachable external CA corresponding to the next CA certificate")
+		}
+	}
+
+	return nil
+}
+
+// If the desired cert is the same as the current CA cert, or is a different cert with the same key and subject, then abort
+// any outstanding rotations. The current signing key is replaced with the desired signing key (this could lets us switch
+// between external->internal or internal->external without an actual CA rotation, which is not needed because any leaf
+// cert issued with one CA cert can be validated using the second CA certificate).
+// Returns: (whether or not the root CA has been updated, the new RootCA with aborted root rotation, any errors)
+func maybeAbortRootRotation(
+	ctx context.Context, cluster *api.Cluster, newCARootCA *ca.RootCA, parsedNewCert *x509.Certificate) (bool, *api.RootCA, error) {
+	if bytes.Equal(cluster.RootCA.CACert, cluster.Spec.CAConfig.SigningCACert) {
+		copied := cluster.RootCA.Copy()
+		copied.CAKey = cluster.Spec.CAConfig.SigningCAKey
+		copied.RootRotation = nil
+		copied.LastForcedRotation = cluster.Spec.CAConfig.ForceRotate
+		return true, copied, nil
+	}
+
+	oldCert, err := helpers.ParseCertificatePEM(cluster.RootCA.CACert)
+	if err != nil {
+		log.G(ctx).WithError(err).Error("invalid current root certificate in cluster")
+		return false, nil, grpc.Errorf(codes.Internal, "invalid current root certificate in cluster")
+	}
+	if ca.SameSubjectAndPublicKey(oldCert, parsedNewCert) {
+		return true, &api.RootCA{
+			CACert:     cluster.Spec.CAConfig.SigningCACert,
+			CAKey:      cluster.Spec.CAConfig.SigningCAKey,
+			CACertHash: newCARootCA.Digest.String(),
+			JoinTokens: api.JoinTokens{
+				Worker:  ca.GenerateJoinToken(newCARootCA),
+				Manager: ca.GenerateJoinToken(newCARootCA),
+			},
```

Comment:
After some discussion with @diogomonica, it may not be worth it to check this particular case.  We're optimizing for not having to generate a new cross-signed certificate, but it might simplify the code if we don't try to do this optimization and just do a full root rotation object instead.  How would you feel about removing this check?